Automattic/jetpack
# Markdown Gutenberg block for Jetpack

### Changes proposed in this Pull Request:

#### Description

Creation of a Markdown Gutenberg block. The user will be able to enter Markdown code in Gutenberg, and it will be rendered into HTML in the client. The block will allow creating new content using Markdown, as well as editing existing Markdown blocks.

This PR comes from the previous work done by @nuzzio in the PRs #9357 and https://github.com/WordPress/gutenberg/pull/6491 (thanks!) and addresses #9201. As a summary, a decision was made to perform the Markdown to HTML conversion in the client, which is the purpose of this PR.

The block will have the following features:

1. [x] Adds a content-editable element to input Markdown text.
2. [x] Adds a preview zone to display the HTML rendered from the Markdown code.
4. [x] Live preview: entered Markdown code will be rendered as HTML as the user types it.
3. [x] Doesn't load the old classic markdown support code if the Gutenberg plugin is installed and active.
6. [ ] ~~Basic text editing buttons to make text bold and italic, create links and add images. They should mimic the UX in the RichText Gutenberg editor component.~~
7. [ ] ~~Add a checkbox to allow the user disable the live-preview (for very big markdown sources).~~
5. [ ] ~~A Markdown syntax legend in the block description (for Markdown newbies).~~

The last three features have been moved to PR #9825

The component will be developed following the order specified in the list above.

This PR notes will be updated as it evolves.

### Testing instructions:

Steps to test the Markdown block component:

1. Install the Gutenberg plugin in your WordPress environment.
2. Enable "Write posts or pages in plain-text Markdown syntax" in your Jetpack config page.
3. Create a new post, add a Markdown block and write some markdown code.
4. Save a draft of the post, or publish it.
5. The markdown code should be rendered as HTML in the published post/draft.

Posts can be edited as well.

~~The block Javascript code will be unit tested using the updated test setup in the PR #9687 that allows testing of Javascript code within modules.~~

~~The block development will be test driven so the test suite will grow as the development progresses.~~

Unfortunately, until issue https://github.com/WordPress/gutenberg/issues/3955 is addressed, custom blocks developed outside Gutenberg cannot be unit tested due to the dependencies they have with Gutenberg core blocks and components.

### Proposed changelog entry for your changes:

Added a markdown block for the Gutenberg plugin.

### Checklist:
